### ISSUE TITLE
Classic shell needs to configure itself for the new operating system

### ISSUE THREAD

**Issue by dbdan22** [Open-Shell/Open-Shell-Menu#167]
Hi,

Well I've put off Win 10 Build 1803 as long as I can, but it's going to come in any day now on about 30 some odd PC's.

Been using Classic Shell for a while, want to try Open Shell. We don't allow our users to be Admin.

Just wondering if non-Admin users will get the same message "Classic shell needs to configure itself for the new operating system" and require Admin priv's to proceed when the new build is installed. 

If so are there plans to fix it, and/or is there a workaround?

Thanks.

**Comment by sarthurdev** [comment]
If this is a domain environment, you could use GPO to do the tasks run by the upgrade.

`HKEY_LOCAL_MACHINE\Software\OpenShell\OpenShell\WinVersion` needs updating to the latest Windows version (The code uses the file version of user32.dll).

The upgrade then calls `regsvr32 /s "%windir%\System32\StartMenuHelper64.dll"` (change to StartMenuHelper32.dll for 32-bit OS),

I suppose you could set the WinVersion registry key to something very high so the upgrade check is always false, I'm not sure if the DLL actually needs re-registering on every Windows 10 feature update.

**Comment by dbdan22** [comment]
We have several sites, both domain and workgroup.

**Comment by sarthurdev** [comment]
For the domain PCs it's as simple as above, however for workgroup non-admin users you will probably need to manually run the upgrade.

**Comment by dbdan22** [comment]
> I suppose you could set the WinVersion registry key to something very high so the upgrade check is always false,

Why is there a need for the WinVersion registry key in the first place if you can fake it?

How difficult would it be to write a system service, manual start, that would sit idle most of the time, until Open Shell detected a build change? Then it would kick off the system service to do this.

**Comment by Nodens-** [comment]
A dll once registered, does not need to be registered again after an update. A dll can get unregistered by an AV application though or registry can get corrupted requiring to register again.

 The way I see it, making an educated guess, this mechanism was implemented in order to deal with some kind of issue in a hackish way. Perhaps at some point windows build upgrades were causing the dll to get unregistered somehow (maybe windows defender or other AV product had something to do with it).
 So if this is tested in VM and the dll remains registered after build update, this code should be actually removed as it's actually bad practice to implement fixes not actually required just to cover edge cases. Specially when those require elevation.
Users actually hitting a problem of an unregistered dll due to an edge case should just rerun the installer or in case of large network environments contact the admins which they will have to do anyway since they can't get elevated rights.

**Comment by dbdan22** [comment]
Agreed.

Well please pardon my ignorance for this is all new to me. But the string "Open-Shell needs to configure itself for the new operating system." is assigned resource 3575 and ID IDS_UPGRADE_ERROR, which is only referenced once in StartMenu.cpp. Now as to what the developer is trying to do leading up to this message, I'll have to stare at this cross-eyed for a few more hours to figure that out...

**Comment by dbdan22** [comment]
> The way I see it, making an educated guess, this mechanism was implemented in order to deal with some kind of issue in a hackish way.

Actually, Ivo addressed this in a post here:
[http://www.classicshell.net/forum/viewtopic.php?f=7&t=8082](url)

`The helper DLL does more than just update it's COM registration.
It hooks into Explorer in a special way to make it possible for the start menu to launch as soon as you log in, not after 10 seconds (like other auto-start programs).
That registration is lost between Windows updates.`

So for this to work for non-admin users, this would have to be done in a system service.

**Comment by Nodens-** [comment]
Ah, I see. Yes. This should indeed be implemented as a service.

**Comment by dbdan22** [comment]
Can we have this marked as a future enhancement? Thanks.

**Comment by sarthurdev** [comment]
You could in the mean time push a group policy start up script to register the DLL silently, or use task scheduler.

How did you install Open Shell on all these machines in the first place?

**Comment by dbdan22** [comment]
> You could in the mean time push a group policy start up script to register the DLL silently, or use task scheduler.

As stated previously, most PC's are workstation, not domain.
And as also stated previously, it's more than just registering a DLL. Open Shell has to hook itself into Explorer. See Ivo's comment http://www.classicshell.net/forum/viewtopic.php?f=7&t=8082 
What code am I supposed to execute to do this? I haven't found it in the code yet, not that familiar with C++.

> How did you install Open Shell on all these machines in the first place?

One at a time, as the PC's were rolled out, over several years. Now I have to go running around to several sites for 10 seconds of work, to log in as Admin and "configure Open Shell for the new operating system", which happens whenever Windows decides to install a new build, a process which I have no control over (other than to delay it by 365 days).

The process of writing a system service is rather straightforward, there are a number of templates for it. Here's one: https://code.msdn.microsoft.com/windowsapps/CppWindowsService-cacf4948 . You cut and paste the code from Open Shell to register the DLL and hook itself into Explorer, and you're done.

So basically the service can get installed with manual start. When Open Shell detects that a new build came in, instead of doing the work itself, it just kicks off the system service and waits a few seconds. So you don't need group policy or a scheduled task to do something every single time you turn on the computer and/or log in, when 99% of the time there will be nothing to do. 

A program should not be crippled for one class of users when it is fairly simple to fix this.

**Comment by ge0rdi** [comment]
> What code am I supposed to execute to do this?

You just need to run:
`"C:\Program Files\Open-Shell\StartMenu.exe" -upgrade`

This is what `Open-Shell Menu` does on startup when it detects that OS version changed.

**Comment by sarthurdev** [comment]
> And as also stated previously, it's more than just registering a DLL. Open Shell has to hook itself into Explorer. See Ivo's comment http://www.classicshell.net/forum/viewtopic.php?f=7&t=8082

The code that is run when the helper DLL is registered is the what applies the explorer hook and updates the WInVersion in registry so the "Configure Open Shell..." popup does not appear.

See here: https://github.com/Open-Shell/Open-Shell-Menu/blob/master/Src/StartMenu/StartMenuHelper/StartMenuHelper.cpp#L317

**Comment by dbdan22** [comment]
This might be a little easier. What if this were done via Scheduled Task?

The equivalent would be along these lines:
https://www.digitalcitizen.life/use-task-scheduler-launch-programs-without-uac-prompts
Making sure that "allow task to be run on demand" and "run with the highest privileges" are selected,

except that Open_shell can create the task programmatically on installation, using this as a template
https://docs.microsoft.com/en-us/windows/desktop/taskschd/weekly-trigger-example--c---

and trigger the task when it detects a new build, using this as a template
https://docs.microsoft.com/en-us/windows/desktop/taskschd/c-c-code-example-starting-a-task

??

**Comment by NomDeMorte** [comment]
I'm going to sound like a rude bastard here so I want you to know that it comes from a good place and is intended to, and will, if you accept it, help you. Please, see past my bastardly tone and see that if you listen to what I'm telling you, everyone wins.


There should be no need whatsoever to change an application in order to deploy it. Learn to deploy it properly.

> As stated previously, most PC's are workstation, not domain.
So what? You can still push policies to workgroup PCs. If you can push anything to them, policies are among those things.

>  which happens whenever Windows decides to install a new build, a process which I have no control over (other than to delay it by 365 days).
Wrong. Learn to configure windows update. There's this myth that win10 prevents users controlling updates. You can still select individual updates just as you use dot be able to in previous versions. You just need to read the docs to find out how.

> A program should not be crippled for one class of users when it is fairly simple to fix this.
A program's development resources should not be wasted in catering for the inadequacies of inexperienced sysadmins.


Again, I know, I sound like a total bastard here. I could spend hours carefully curating this post so as not to hurt your feelings but hopefully you'll just nut up and realise that you're crippling the development resources of this tool by relying upon them to compensate for your lack of experience and knowledge of the tools you're using.

This is not something I say in order to put you down - on the contrary. It's something I say that, hopefully, you will take this constructive criticism, take some time and learn some new skills that will both save this project development time, and expand your skills and career prospects.

**Comment by Kudalufi** [comment]
I have **never** made a meta-post on GitHub, but that requires some correction.

> I'm going to sound like a rude bastard here
> There's this myth that ... you just need to read the docs to find out how.

Translation: _"I can't be bothered to actually help you or the others in this community with the posted problem, even though I know (or claim to know) many if not all of the solutions.  I am going to make vague references to esoteric and arcane documentation that validates my claims, and pat myself on the back afterwards for administering tough love."_

> I could spend hours carefully curating this post so as not to hurt your feelings but hopefully you'll just nut up

How about five minutes for the post's (premeditated) rudeness, half an hour to make it helpful, and the rest of those quoted long hours spent curating and improving yourself so that helping people actually seems like the natural thing to do.

> ... and realise that you're crippling the development resources of this tool by....

- by dangling intentionally malformed non-help in a forum that is specifically designed for the sharing of solutions
- by wasting the development resources (which is all of us) of this tool dealing with your phallic waving

What a missed opportunity.  You had an awesome opportunity to offer some guidance to not only the original poster here, but everyone else who came after.  In the future, lead, follow, or get out of the way.

**Comment by ibuprophen1** [comment]
Can I please get some clarification on this? 

I apologize for my ignorance but, So far it looks like this issue began with someone asking about this development to be revamped (in so many words) and, eventually, it turned into a type of debate (with some choice verbiage) over something off topic. 

Thank you very much for your time and understanding with this question for clarification. :-) 


~Ibuprophen

**Comment by Kudalufi** [comment]
My apologies.  I didn't intend to hijack the whole conversation.  There was no private way to communicate with NomDeMorte or any email address listed on his profile.

You can safely assume the issue is about development tailoring towards Windows 10. :)

**Comment by ibuprophen1** [comment]
@Kudalufi, No Worries! :-) 

There's a handful of issues here that are either off topic and/or duplicates (in one way or another). 

I believe that revamping this development by gearing it towards Windows 10 specifically could be more dangerous to those who doesn't use Windows 10 and open up the Gates of Hades with issues being reported. 

There's other issues reported here that are Windows 10 related that needs to be addressed and, ultimately, improve the compatibility over time. 

I personally believe that this issue is more of a question than an issue myself. 

Unless further objected to, I believe that this issue should be noted and closed. I just don't want to jump the gun and close it without getting some clarification regarding this being an issue or a question that has been answered (or already asked elsewhere). 


~Ibuprophen

**Comment by ibuprophen1** [comment]
Has this issue been resolved @dbdan22? 

I'm asking you since this issue was opened by yourself and guidance was provided. 



~Ibuprophen

**Comment by dbdan22** [comment]
Has this issue been resolved? Well, actually, no. In my book, any application that requires admin priv's to function properly is a poorly designed application, especially when this is fairly easy to fix. But close it if you must.

Oh, and NomDeMorte? No offense taken, 'cause [I too can be a rude bastard](https://www.youtube.com/watch?v=VW5TWYbgMGQ). But I'll not stoop to your level. Thanks for the tips.

**Comment by ibuprophen1** [comment]
I'm not stating that it's not happening to you but, I personally cannot reproduce this myself on my Windows 10 Laptop. 



~Ibuprophen

**Comment by dbdan22** [comment]
That's kinda ironic, 'cause at this very moment I'm sitting around waiting for a build 1803 update to finish so I can log into Admin so it can configure itself. This particular laptop is still running Classic Shell tho... perhaps I should have updated it to Open Shell before the update came in. 20/20 hindsight again

**Comment by ibuprophen1** [comment]
Just as a suggestion... 

Maybe you can get a fresh start (so to speak... LOL!) if you logged in as an Administrator, uninstalled the Classic Shell, reboot and log in to your default profile then install Open Shell. 

If you have further troubles regarding the Administrative capabilities, try the following... 

1) Open the Run Dialog.

2) Enter (without quotes) "compmgmt.msc /s" OR "compmgmt.msc".

3) Expand the "Local Users and Groups" (on the Left) and select "Users".

4) Double Click your Profile Name and select the "Member Of" (top) tab.

5) See if it reflects the image below...

[Example-Screenshot](https://user-images.githubusercontent.com/18060271/58653770-6dee5100-82e4-11e9-9e34-f8658328209f.png)

... if not, add the "Administrator" to it.

6) Afterwards, Reboot your PC and log in to your Profile.

I can't state that it will definately work but, it should (hopefully) work.

Good Luck!


~Ibuprophen

**Comment by ibuprophen1** [comment]
I'm going to close this issue as it seems to be more likely a Windows issue rather than an Open Shell issue. 


~Ibuprophen

**Comment by Nodens-** [comment]
I have no idea why this was marked as Windows bug. This is clearly a design issue with Open Shell.
Ideally when it comes to UX, the reconfiguration process should be moved to a system service and hence drop the elevated rights requirement after windows buils updates.
This should clearly be marked as Enhancement and remain open.

**Comment by ge0rdi** [comment]
I'd also not call this Windows bug.
It was already written that `OpenShell` needs to re-register `StartMenuHelper` after OS upgrade. And this needs administrator rights.

Currently `OpenShell` runs this when it detects that OS version changed:
`"C:\Program Files\Open-Shell\StartMenu.exe" -upgrade`

It basically does just:
`regsvr32 /s "%windir%\System32\StartMenuHelper64.dll"`

We need to investigate whether that's still the case with recent Windows 10 versions. Maybe it is no longer needed and it can be removed (which will elegantly solve this issue).

If there is still a need for the re-registration, we need to at least add possibility to run `-upgrade` switch silently (currently there is message box shown at the end).
This way it will be possible to create scheduled task that will run upgrade task with administrator rights on every boot.

Eventually we can add such task automatically during installation (this is to be considered).

**Comment by ibuprophen1** [comment]
Oh... I See... My apologies @ge0rdi! 👍 

I had tried to duplicate this in a few ways and I was able to experience the same/similar issue (I think). 

It was duplicated by the default windows profile not having the appropriate Administrative Permissions. 

I took care of this by logging in under the Administrators profile and adding the Administrator permissions to the default profile and making sure that no other conflicting permissions were there. 

I hope I had explained this okay via text... 


~Ibuprophen

**Comment by ge0rdi** [comment]
> We need to investigate whether that's still the case with recent Windows 10 versions.

It seems that the issue is still there :(
I've tried to upgrade Win10 1809 to 1903 with modified `OpenShell` (that doesn't handle OS upgrade at all). And apparently this `StartMenuHelper` registration was lost.

After reboot there is standard Windows start menu shown. Then after cca 10 seconds it is switched to `Open-Shell` start menu.
This is exactly what Ivo described here:
http://www.classicshell.net/forum/viewtopic.php?f=7&t=8082#p34821

**Comment by gbjbaanb** [comment]
It seems to me that a better approach would be to commit the modified version that does not do any upgrade checking, and move that upgrade-checking code into a service (or other run-at-system-boot executable) that can perform the upgrade externally to the actual menu replacement.

**Comment by Growlertiger** [comment]
I am still using the final version of Classic Shell. I just encvountered a new issue with updating. I don't know if this would have happened with Open Shell, but I thought I should tell you.
I just upgraded one PC to 1903 and there was no problem. I got the usual message "Classic shell needs to configure itself for the new operating system" and all was fine.
But I then upgraded another PC which was I thought set up the same way, to 1903. This time instead of getting that message I got a large red box I have not seen before, giving the command string with the -upgrade and saying the execution was blocked. It mentioned it not being signed. It did not ask me to approve as an admin, it simply blocked it.
I got it to run by opening a command window as Administrator, going to the folder in Program Files, and manually running the command with -upgrade parameter. This then gave the usual message "Classic shell needs to configure itself for the new operating system" and all was then fine.
Perhaps this will happen to others upgrading to 1903.

**Comment by ge0rdi** [comment]
@gbjbaanb 
I don't think there is a need for service just because of this OS upgrade handling.

What I've prepared in the meantime is possibility to run:
`%ProgramFiles%\Open-Shell\StartMenu.exe -upgrade -silent`
That will do all necessary things if OS version changed, without any user interaction.

This can be then easily run as scheduled task with system rights on every boot.
Admins can easily add such task on machines they manage.
See #176.

In the future we may eventually add option to create such task into installer.

**Comment by ibuprophen1** [comment]
@ge0rdi, I don't believe, personally, that a silent install would be something that the majority would want as it could, potentially, open up allot of potential issues in the end. 

Silent installation is fine for some software and such but, I don't feel that this would have any beneficial value to the Open Shell software. 

One reason would be that i've personally seen software with silent installations that, if errors & hangs (being stuck) somewhere in the process, not being known to the individual. 

Also, I've seen options that are provided, to the individual to choose from, not being provided/available as its running in the background. 

This is just my opinion, suggestion, recommendation, etc... 😱 


~Ibuprophen

**Comment by ge0rdi** [comment]
Not sure what do you mean exactly.
But this is not about silent installation.

What I wrote (and prepared PR for) is about the possibility to do OS upgrade related tasks in a way that don't require user interaction.

**Comment by gbjbaanb** [comment]
any way will work - a service is one way, silent upgrade task run on boot is another.  Technically a service is better as you install it with admin rights and it can then do what it likes - typically startup, check, then stop itself. That's the main reason for using one, just another way of getting the sam,e work done - though I think it can be set to run before any scheduled boot tasks, which could be important.

**Comment by ibuprophen1** [comment]
I see... @ge0rdi! 👍 

@gbjbaanb, I do agree with implementing the Elevated Permissions for the software. 

I just don't want anyone to think that, if it's not happening to me then it's not happening to anyone else. 

I do understand that it's something that others are experiencing and support it as well. 


~Ibuprophen

**Comment by ge0rdi** [comment]
Latest nightly build [4.4.138](https://ci.appveyor.com/project/passionate-coder/open-shell-menu/builds/25153094/artifacts) supports

`%ProgramFiles%\Open-Shell\StartMenu.exe -upgrade -silent`

That will perform all necessary tasks without any user interaction.

**Comment by ge0rdi** [comment]
I have created wiki page regarding this issue:
https://github.com/Open-Shell/Open-Shell-Menu/wiki/OS-upgrade-without-user-interaction

**Comment by jdunn0** [comment]
Reply to the comment made by [NomDeMorte on 2019-05-28](https://github.com/Open-Shell/Open-Shell-Menu/issues/167#issuecomment-496459730):

Based on the first part of your comment, it seems you don't understand what the actual problem here is.
Someone installs Open-Shell, and later installs a new feature update for Windows 10.
The Open-Shell version is still the exact same as before but part of it was broken in the process of upgrading the OS.
Open-Shell is not being installed as an update or a new version, that was already done, it's the Windows Update that causes the installed version of Open-Shell to no longer work properly and need a repair.
Open-Shell auto starts on login to the OS like it normally would, notices that part was broken and offers to repair the part that was broken.
The problem is that while Open-Shell doesn't require Administrative Privileges to run, they are needed for the repair process.
There isn't any problem with the initial deployment of Open-Shell, that was fine, it was an OS update that required Open-Shell to run the repair step.

> Wrong. Learn to configure windows update. There's this myth that win10 prevents users controlling updates. You can still select individual updates just as you used to be able to in previous versions. You just need to read the docs to find out how.

As someone who has used Windows 10 for a while now and used the update system, I can say with 100% certainty that you incorrect here.
Windows 10 only gives you two options, install all updates or don't use Windows Update at all (which isn't an easy thing to do but possible).
The ability to select individual updates is gone from Windows 10, there is no way to do that at all.
I have no clue why you think this can be done as it has never been possible in Windows 10.

> A program's development resources should not be wasted in catering for the inadequacies of inexperienced sysadmins.

That is not the case here, the sysadmin has deployed the software and applies Windows Updates, there is no evidence to support that argument.
The fact that those Windows Updates cause Open-Shell to be broken until a repair is done by an Administrative user is a problem that shouldn't exist but it does and should be addressed.

Reply to the comment by [dbdan22 on 2019-05-30](https://github.com/Open-Shell/Open-Shell-Menu/issues/167#issuecomment-497276935)
> Has this issue been resolved? Well, actually, no. In my book, any application that requires admin priv's to function properly is a poorly designed application, especially when this is fairly easy to fix. But close it if you must.

Actually Open-Shell doesn't require Administrator Privileges to function properly.
It's only the repair step that is needed because a Windows Update broke part of it that needs Administrative Privileges to repair the program.
It's not really poor design of the program, there are limited options on what can be done with this issue.

Ideally Microsoft should add a Start Menu Replacement API to Windows 10 so that Open-Shell doesn't have to use deep shell integration that is broken by a Windows Update.

As that is not the case, what should be done is the figure out a way to get a Scheduled Task to run after an upgrade was installed to run the upgrade task with Administrative Privileges.
There are two ways I think a scheduled task (created by the Open-Shell installer) could work:
1. Have the task run a process on system startup that checks if an upgrade is done and run the repair code if that is the case.
The system startup should mean that it runs before a user is logged in to an account but due to timing and automatic login that wouldn't always happen.
2. Have the task run on a trigger to run the repair step.
You can create a task that runs when a specific event happens.
If there is an event in the event log that always appears after a feature update that could be used as a trigger then the task would be run when that happened.
This option is tricky though, I have no idea if such an event is logged and if that event can used as a trigger for the task..
3. Create a task to do the repair with no triggers, allow it to be run on demand and have the non-admin process start the task which will run with Administrative Permissions.
This method was mentioned in a comment by [dbdan22 on 2019-05-11](https://github.com/Open-Shell/Open-Shell-Menu/issues/167#issuecomment-491546572).

It's hard to say how easy option 2 or option 3 would be to implement, the only real problem with option 1 is that a process always runs when the computer boots up regardless of if it needs to be run but it would at least not run every time a user logs in.

In reply to the comment by [Nodens on 2019-06-04](https://github.com/Open-Shell/Open-Shell-Menu/issues/167#issuecomment-498882629):
> Ideally when it comes to UX, the reconfiguration process should be moved to a system service and hence drop the elevated rights requirement after windows buils updates.

A service would work too but a process run by task scheduler would only run once while a service would likely run all the time and that is not necessary.

**Comment by ge0rdi** [comment]
> Have the task run a process on system startup that checks if an upgrade is done and run the repair code if that is the case.

This is already possible with latest nighly builds and it is recommended (for now):
https://github.com/Open-Shell/Open-Shell-Menu/wiki/OS-upgrade-without-user-interaction

Though one has to create such scheduled task by himself (just once after `Open-Shell` installation).

**Comment by Nodens-** [comment]
> A service would work too but a process run by task scheduler would only run once while a service would likely run all the time and that is not necessary.

No. You are wrong on this one. The service would start on boot, apply the hook if needed and terminate itself. It is actually the  best option for 2 reasons.

1) Service executes prior to scheduled tasks, really early, so you avoid any possible race condition (eg user logging in before the hook has been applied).
2) Does not depend on the task scheduler which is an extra point of failure and is also not affected by scheduled task congestion which may cause user being logged in before the hook is applied.

Since we're talking deployment in large networks with x number of workstations, the actual cases of any of the above happening is not that rare.
Of course the current task scheduler based solution is acceptable and better than having an admin jumping through hoops. Better to have to reconfigure the odd system here and there or even telling the user to reboot than having to deal with hundreds of systems heh.

**Comment by NomDeMorte** [comment]
> The problem is that while Open-Shell doesn't require Administrative Privileges to run, they are needed for the repair process.

So, don't update as the user, and/or don't use the repair process. There are a bazillion apps out there that have something that requires admin privileges, where the user doesn't get admin. How do you think they do it? They don't get the apps changed.

> As someone who has used Windows 10 for a while now and used the update system, I can say with 100% certainty that you incorrect here.
> I have no clue why you think this can be done as it has never been possible in Windows 10.

I'll just leave this here to get you started: https://duckduckgo.com/?q=AUOptions

> The ability to select individual updates is gone from Windows 10, there is no way to do that at all.

https://www.catalog.update.microsoft.com/home.aspx
https://duckduckgo.com/?q=wushowhide.diagcab

There's much more to be said on this topic, but this thread is not "How to sysadmin Win10". That should get you started.

> sysadmin has deployed the software and applies Windows Updates, there is no evidence to support that argument.
> The fact that those Windows Updates cause Open-Shell to be broken until a repair is done by an Administrative user is a problem that shouldn't exist but it does and should be addressed.

If sysadmin rolls out an update to one component without the update to another component which is broken by the first, sysadmin is doing it wrong.

If sysadmin strips users of admin rights, without a mechanism to perform admin tasks, sysadmin is doing it wrong.


I'm really not interested in a lengthy discussion about this. This is the sysadmin's issue to fix, not the application developer's. What would you do if this application was made by a non-open-source team that wouldn't take user requests on board? Do that.

**Comment by dbdan22** [comment]
Services do not have to run at all, they can be set up as Manual, to be run only on demand. 

When Open Shell detects that the service is necessary, it would start the service, which would do the configuration, and then terminate. The only thing is, you have to set the permissions on the service properly to allow a non-admin to start it: http://woshub.com/set-permissions-on-windows-service/. 

With a service, open shell does everything, there's no need to create a scheduled task, and there's no need for a scheduled task to run at every boot when 99.9% of the time it won't do anything.

**Comment by dbdan22** [comment]
[A basic Windows service in C++ (CppWindowsService)](https://code.msdn.microsoft.com/windowsapps/CppWindowsService-cacf4948)

   // Service start options. 
   #define SERVICE_START_TYPE       SERVICE_DEMAND_START 

[Starting Services on Demand](https://docs.microsoft.com/en-us/windows/win32/services/starting-services-on-demand)

Service gets created and runs under System. It does not do anything until called upon by Open Shell. 

This is not difficult.

**Comment by dbdan22** [comment]
So have we reached a consensus on how to proceed?

**Comment by ge0rdi** [comment]
@dbdan22 
I personally have no objections against on-demand service. It can be hosted within `StartMenu.exe` and started only when OS change was detected.

Other similar approach could be to create scheduled task in installer, but don't run it on each boot. Instead the task can be also triggered manually (when OS change was detected).

Though I'd suggest to keep this optional (installer option, or simply just MSI property that can be set from command line).

Feel free to experiment with it and submit PR. I'm willing to do review.

**Comment by dbdan22** [comment]
My coding skills are, um, a bit rusty. It would take me awhile to come up to speed. I'm an IT tech by trade.

**Comment by jdunn0** [comment]
> No. You are wrong on this one. The service would start on boot, apply the hook if needed and terminate itself.

Yeah, you are right.
I didn't think of the things you pointed out.

> So, don't update as the user, and/or don't use the repair process. There are a bazillion apps out there that have something that requires admin privileges, where the user doesn't get admin. How do you think they do it? They don't get the apps changed.

Open-Shell is not like other apps, it hooks into the Windows Shell so that the start button opens the Open-Shell Start Menu.

When a new Windows 10 Feature Update is installed, this hook is no longer operational.
The updates are installed by the system using the Automatic Updates feature of Windows 10 so it's already running with Administrator Privileges.

Open-Shell runs without Administrator Privileges but due to that hook not being operational, it has to ask for Administrative Privileges to do what the Open-Shell installer would do when the hook is first installed so it's almost like doing a reinstall of the program so it works again.
All install, repair and uninstall tasks for most apps need Administrative Privileges to work even though the app doesn't need them.

There are apps out there that put all install, repair and uninstall tasks in a service so that those tasks can be done by users without Administrative Privileges.

Examples:
Mozilla Firefox and other Mozilla products install the "Mozilla Maintenance Service" which handles doing updates for the user so the user doesn't see elevation prompts every time there is an update.
Google Chrome installs Google Update for the same reason.
The services for both only run when needed, they don't run all the time.
There is also services for Adobe products, Valve's Steam, EA's Origin, GOG Galaxy and others.
Not to mention the "Windows Update" service that let's users see and control (if their Windows edition allows it) updates without needing to be an Administrator to do so.
Also DisplayFusion, Stardock Groupy, Synergy, and others use services so the programs can run without Administrative Privileges but still be able to do things that need them.

In short: Most apps that regularly need Administrative Privileges install a service to handle that.

> https://www.catalog.update.microsoft.com/home.aspx

That is not a solution to the problem of not being able to select updates.
Sure, you can download just the updates you want from the Update Catalog Site but the Windows 10 Automatic Updates feature will just download any updates you didn't pick anyway unless you disable the Automatic Updates feature entirely (which isn't always possible in Windows 10).
Also the Automatic Updates feature is more bandwidth efficient then the Update Catalog is, when you download an update from the Update catalog you get the entire update in one file, sometimes though you won't need everything in that update so the Automatic Updates feature will only download the parts it knows are needed for that system.

> If sysadmin rolls out an update to one component without the update to another component which is broken by the first, sysadmin is doing it wrong.

In most cases, it's the Windows 10 Automatic Updates feature installing a new Feature Update, not the sysadmin/user so the user without Administrator Privileges can't do anything without them to repair a program broken by that update.

In a way, Windows 10 itself is acting as the sysadmin of the Windows 10 system in that case.

> What would you do if this application was made by a non-open-source team that wouldn't take user requests on board?

Either find a workaround or stop using that application.
However as shown above when it was deemed necessary, many applications will have services made by the application developers in the first place to avoid unnecessary prompts so it won't normally be a problem anyway.

**Comment by NomDeMorte** [comment]
I understand the nature of the problem and the hooks, and while it is unusual, that is not unlike every other app. I've had to package apps that do just this. That's why I'm explaining how to deal with it appropriately. I'm not new to this. This first time I had to deal with this was just over 20 years ago, preparing for Y2K, deploying an app which modified windows' behaviour to communicate with factory automation hardware.

Automatic updates can be prevented, I posted links about that, and in the case of a managed environment, they should be, because the sysadmin should be rolling them out only after testing.

Users should not be allowed to apply software updates in a managed environment, again because updates should be tested prior - this is equally true for applications and OS updates.

Applying updates selectively is entirely possible by selecting them from the catalog and because automatic updates are prevented. Of course, there are more appropriate methods, I'm trying to keep it brief to avoid this thread turning into sysadmin 101.

Those automatic update services are there for end-user convenience and do not apply to a managed environment where they will always be disabled to ensure uniform and controlled management of the platform.

If there is a desire to expend development resources on providing such a service as an end-user convenience then that is one matter, but implementing it to avoid proper administration of a managed environment, which is what this thread is about, is a different matter.

**Comment by Nodens-** [comment]
> If there is a desire to expend development resources on providing such a service as an end-user convenience then that is one matter, but implementing it to avoid proper administration of a managed environment, which is what this thread is about, is a different matter.

Administration practices have nothing to do with it, really. The fact that there are ways around it and that there is a lot of software out there that is written with bad coding practices is also irrelevant.
This is a valid concern. The application should not require it and it should handle any function requiring elevation transparently, since it is possible. Otherwise it is just bad design. And I can say that as a software engineer, working with native code for more than 25 years.

I am not affected by this issue as I don't administer windows systems but when time allows, as I'm currently drowning in work, and if someone doesn't beat me to it, I'll submit a PR myself as this is the most robust solution that avoids other problems as well (the ones mentioned in my previous post).

**Comment by NomDeMorte** [comment]
> Administration practices have nothing to do with it, really
I mean that's literally the subject being discussed here. This request is not being made for reasons of streamlining the end user experience.

> This is a valid concern.
Yes it is, I never suggested otherwise. Another valid concern is that this project is light on resources (as in, [not much dev time](https://github.com/Open-Shell/Open-Shell-Menu/commits/master)) and while the update process being mildly inconvenient is a valid concern, the software is entirely functional as it stands, and it would be a waste of the limited resources given to the project used by many, to make allowances for lone a sysadmin who can't figure out how to deploy apps and updates. 

If you have such development experience then surely your managers have already made you aware of the necessity of appropriate investment of resources. Like I said, if someone wants to do this as a convenience, that's not a problem. It's only a problem if that time is taken from elsewhere in the project, to compensate for one sysadmin who's doing it wrong, at the expense of the health of a project that's been near death too many times for any of our comfort. 

To put it simply, my issue is not the "what", it's the "when". It's not a question of validity, it's a question of priority. This facet of the software, while being somewhat suboptimal, is entirely functional, which can't be said for the entire project. There are bigger fish to fry, here, and not much fire. I feel I've made this quite clear all along, but I understand not having time to parse the whole thread if you're swamped.

If you'd like to resolve all of this by taking on the time investment yourself and are sure there aren't any [higher severity issues](https://github.com/Open-Shell/Open-Shell-Menu/issues) which your available time investment could resolve, kudos and thanks to you. Nobody is going to resist that. What would be a shame, is a dev with 25 years of experience wasting his valuable time fixing something that isn't broken and ignoring things that are.

**Comment by Nodens-** [comment]
Frankly there are no other reproducible issues on the tracker. Only non-reproducible reports which look like user errors (something else wrong with the user's system) and a few feature suggestions that I personally think of much lower priority (of those that I would personally consider for addition but that is not my call anyhow as I'm not part of the team).
This, however, is pretty valid, in my eyes, for the possibility alone that a user may actually log in before the task scheduler runs any related scheduled task as I've already explained. Like I said, the original design was just bad. Hence I'm willing to submit a PR.
Regarding your other comments, not that they have much place in an issue tracker but for the past 16 years or so, I do exclusively consigned/contracted work so I don't have managers. Only clients. But back then, when I did, they would agree on prioritizing this over everything else currently in the tracker (none reproducible reports and feature requests) for one reason alone. Less support calls.

On a last note, you seem pretty adamant and passionate about opposing this, going into great lengths to debate it over, quite aggressively at times too, when the issue has already been accepted by the project team (ge0rdi has already provided a working solution) and like I said I'm willing to submit a PR for the service when time allows, since ge0rdi said he's willing to accept it, so for all purposes the issue is considered set and done.
Since you're not interested in submitting a PR yourself, then there's no point in your participation in this discussion as further debating it, or derailing it into system administration practices, serves zero purpose other than cluttering the tracker and spamming the watchers' mailboxes.

**Comment by NomDeMorte** [comment]
> you seem pretty adamant and passionate about opposing this
I'm just replying to people who don't understand it, so 

>  spamming the watchers' mailboxes.
Is on them. Don't expect to engage me without a reply.

> derailing it into system administration practices
Derailing? That's literally the subject here.

>  no other reproducible issues on the tracker
Then this:

> If you'd like to resolve all of this by taking on the time investment yourself and are sure there aren't any [higher severity issues](https://github.com/Open-Shell/Open-Shell-Menu/issues) which your available time investment could resolve, kudos and thanks to you

Applies.

**Comment by dbdan22** [comment]
When might a version with a system service be available for beta testing, if I may be so bold as to ask? Thanks.

**Comment by leknowlton** [comment]
The issue is real and something that should have been fixed near the program's inception.  A program that becomes so integral to an OS's UI should not require any elevated intervention, after an OS update.  Intervention should only be required during the actual program update process or removal.  Upon login, this program should detect the OS, check if the OS has been updated and fix the hook into Explorer that the **program created**.  It's not the users fault the OS updated and removed the hook, so why punish them by forcing them to manually redo it at an elevated level?   

As it has already been stated: This has been accepted as an issue and solutions that don't sound overly burdensome to create have been proposed that would remedy it.  

Increasing the amount of the program's autonomy from user interaction will increase its relevance/persistence, increase the possibility that people will actually donate some $$ towards its development/maintenance and decrease uninstalls due to having to deal with this issue.

**Comment by NomDeMorte** [comment]
>  It's not the users fault the OS updated and removed the hook
No, it's the administrator's fault. Fortunately, the administrator has the tools to prevent updates, and to replace the hook when they allow the update. 

> The issue is real ... This has been accepted as an issue 
I don't see anywhere that anyone has suggested it wasn't an issue.

Meanwhile, despite the above discussion, there are in fact real issues which are repeatable and long-standing, which cannot be addressed by the administrator doing their job, and require patches to the software.

> solutions that don't sound overly burdensome to create have been proposed that would remedy it.
Totally, so why is everyone so resistive to those solutions? Why the constant nagging argument over prioritising workloads appropriately? I'm growing tired of repeating myself and burning straw men.

> uninstalls due to having to deal with this issue.
Has this happened? Has it happened more than people uninstalling over things they can't fix at the click of a button, or by managing their network as though it's literally their job? Where are these statistics?

**Comment by leknowlton** [comment]
How nice NomDeMorte that you decide to continuously respond with "your way or the highway", when others have provided possible solutions to the upgrade issue that alleviate end user interaction and admins from having to add complexity to their networks.

**Comment by leknowlton** [comment]
I believe I tried running said silent install upon login and still got UAC.

**Comment by ge0rdi** [comment]
@leknowlton 
Have you followed [OS upgrade without user interaction](https://github.com/Open-Shell/Open-Shell-Menu/wiki/OS-upgrade-without-user-interaction) guide?

You need to run the _upgrade_ task as elevated admin. Then there should be definitely no UAC prompt.
